authorization/src/main/kotlin/org/modelix/authorization/ModelixJWTUtil.kt

```diff
@@ -0,0 +1,312 @@
+        return RSAKey.Builder(key.toRSAKey()).algorithm(JWSAlgorithm.RS256).build()
+    }
+
+    private fun ensureKeyId(key: JWK): JWK {
+        if (key.keyID != null) return key
+
+        val rsaKey = key.toRSAKey()
+        val digest = MessageDigest.getInstance("SHA-256")
+        digest.update(rsaKey.modulus.decode())
+        digest.update(0)
+        digest.update(rsaKey.publicExponent.decode())
+        val keyId = Base64.getUrlEncoder().withoutPadding().encodeToString(digest.digest())
+        return RSAKey.Builder(rsaKey).keyID(keyId).build()
+    }
+
+    fun loadJwkFile(fileContent: String): JWK {
+        return JWK.parse(fileContent).also { loadJwk(it) }
+    }
+
+    private fun loadJwk(key: JWK) {
+        if (key.isPrivate) {
+            setRSAPrivateKey(key)
+        } else {
+            addPublicKey(key)
+        }
+    }
+
+    fun verifyToken(token: String) {
+        DefaultJWTProcessor<SecurityContext>().also { processor ->
+            val keySelectors: List<JWSKeySelector<SecurityContext>> = hmacKeys.map { it.toPair() }.map {
+                SingleKeyJWSKeySelector<SecurityContext>(it.first, SecretKeySpec(it.second, it.first.name))
+            } + jwksUrls.map {
+                val client = this.ktorClient
+                if (client == null) {
+                    JWSAlgorithmFamilyJWSKeySelector.fromJWKSetURL<SecurityContext>(it)
+                } else {
+                    JWSAlgorithmFamilyJWSKeySelector.fromJWKSource<SecurityContext>(RemoteJWKSet(it, KtorResourceRetriever(client)))
+                }
+            } + rsaPublicKeys.map {
+                JWSAlgorithmFamilyJWSKeySelector.fromJWKSource<SecurityContext>(ImmutableJWKSet(JWKSet(it.toPublicJWK())))
+            }
+
+            processor.jwsKeySelector = if (keySelectors.size == 1) keySelectors.single() else CompositeJWSKeySelector(keySelectors)
+
+            val expectedKeyId = this.expectedKeyId
+            if (expectedKeyId != null) {
+                processor.jwsVerifierFactory = object : DefaultJWSVerifierFactory() {
+                    override fun createJWSVerifier(header: JWSHeader, key: Key): JWSVerifier {
+                        if (header.keyID != expectedKeyId) {
+                            throw BadJOSEException("Invalid key ID. [expected=$expectedKeyId, actual=${header.keyID}]")
+                        }
+                        return super.createJWSVerifier(header, key)
+                    }
+                }
+            }
+        }.process(JWTParser.parse(token), null)
+    }
+
+    class TokenBuilder(private val builder: JWTClaimsSet.Builder) {
+        val claimSetBuilder: JWTClaimsSet.Builder get() = builder
+        fun claim(name: String, value: String) {
+            builder.claim(name, value)
+        }
+    }
+}
+
+class KtorResourceRetriever(val client: HttpClient) : AbstractRestrictedResourceRetriever(1000, 1000, 0) {
+    override fun retrieveResource(url: URL): Resource? {
+        return runBlocking {
```

Comment:
The `JWTProcessor` instance and all the key sources are now reused which allows `RemoteJWKSet` to cache the keys instead of having to download them for each request.